src/main/java/com/bobocode/orm/context/PersistenceContext.java

```diff
@@ -0,0 +1,76 @@
+package com.bobocode.orm.context;
+
+import static com.bobocode.orm.util.EntityUtils.extractId;
+import static com.bobocode.orm.util.EntityUtils.toSnapshot;
+
+import com.bobocode.orm.util.EntityKey;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class PersistenceContext {
+
+  private final Map<EntityKey<?>, Object> entitiesCache = new HashMap<>();
+  private final Map<EntityKey<?>, List<?>> entitiesSnapshots = new HashMap<>();
+
+  public boolean contains(@NonNull Object entity) {
+    EntityKey<?> key = EntityKey.of(entity.getClass(), extractId(entity));
+    return entitiesCache.containsKey(key);
+  }
+
+  public boolean contains(@NonNull EntityKey<?> key) {
+    return entitiesCache.containsKey(key);
+  }
+
+  public <T> T getEntity(@NonNull EntityKey<T> key) {
+    return key.getType().cast(entitiesCache.get(key));
+  }
+
+  public void addEntity(@NonNull Object entity) {
+    EntityKey<?> key = EntityKey.of(entity.getClass(), extractId(entity));
+
+    entitiesCache.put(key, entity);
+
+    List<?> snapshot = toSnapshot(entity);
+    log.trace("Created snapshot for {} entity: {}", entity.getClass().getSimpleName(), snapshot);
+    entitiesSnapshots.put(key, snapshot);
+  }
+
+  public List<?> getDirtyEntities() {
```

Comment:
It is not a good practice to use wildcards in the return type, especially in public methods. I think in this method, as well as in the toSnapshot() method, you can return just List of Objects, because in any case you put Objects in snapshot lists since the Field.get() method returns an Object type (line 53).

Reply:
Agreed, will address it a bit later.
Thx for the feedback.

Reply:
fixed